File: engine/signals.py
#!/usr/bin/env python3
"""
SkillGod Layer 2 — Local Signal Tracking.

Signals are opt-in (default OFF). All data stays on-device in SQLite.
When SKILLGOD_API is set, flush_signals() sends to the Railway backend.

Kinds:
  accept  — skill fired, no rework detected  → score 1.0
  rework  — skill fired, rework detected      → score max(0, 1 - count*0.25)
"""

import json
import os
import re
import sqlite3
from datetime import datetime
from pathlib import Path

ROOT    = Path(__file__).parent.parent
DB_PATH = ROOT / "db" / "skillgod.db"

ENABLED_KEY = "signals_enabled"


# ---------------------------------------------------------------------------
# Schema
# ---------------------------------------------------------------------------

def _get_db() -> sqlite3.Connection:
    DB_PATH.parent.mkdir(parents=True, exist_ok=True)
    conn = sqlite3.connect(str(DB_PATH))
    conn.row_factory = sqlite3.Row
    conn.executescript("""
        CREATE TABLE IF NOT EXISTS signals (
            id          INTEGER PRIMARY KEY AUTOINCREMENT,
            skill_id    TEXT    NOT NULL,
            skill_name  TEXT    NOT NULL,
            session_id  TEXT    NOT NULL,
            kind        TEXT    NOT NULL CHECK(kind IN ('accept','rework','learned')),
            rework_count INTEGER DEFAULT 0,
            score       REAL    NOT NULL,
            created_at  TEXT    NOT NULL
        );
        CREATE INDEX IF NOT EXISTS idx_signals_skill  ON signals(skill_id);
        CREATE INDEX IF NOT EXISTS idx_signals_kind   ON signals(kind);

        CREATE TABLE IF NOT EXISTS kv (
            key   TEXT PRIMARY KEY,
            value TEXT
        );
    """)
    conn.commit()
    return conn


# ---------------------------------------------------------------------------
# Opt-in control
# ---------------------------------------------------------------------------

def is_enabled() -> bool:
    """Signals are opt-in. Returns True if user has enabled them."""
    try:
        conn = _get_db()
        row  = conn.execute(
            "SELECT value FROM kv WHERE key=?", (ENABLED_KEY,)
        ).fetchone()
        conn.close()
        return row is not None and row["value"] == "1"
    except Exception:
        return False


def enable() -> None:
    conn = _get_db()
    conn.execute(
        "INSERT OR REPLACE INTO kv(key,value) VALUES(?,?)",
        (ENABLED_KEY, "1")
    )
    conn.commit()
    conn.close()


def disable() -> None:
    conn = _get_db()
    conn.execute(
        "INSERT OR REPLACE INTO kv(key,value) VALUES(?,?)",
        (ENABLED_KEY, "0")
    )
    conn.commit()
    conn.close()


# ---------------------------------------------------------------------------
# Recording
# ---------------------------------------------------------------------------

def record_no_rework(skill_id: str, skill_name: str, session_id: str) -> None:
    """Record that a skill fired and output was accepted without rework."""
    if not is_enabled():
        return
    conn = _get_db()
    conn.execute(
        "INSERT INTO signals(skill_id,skill_name,session_id,kind,rework_count,score,created_at)"
        " VALUES(?,?,?,?,?,?,?)",
        (skill_id, skill_name, session_id, "accept", 0, 1.0,
         datetime.now().isoformat())
    )
    conn.commit()
    conn.close()


def record_rework(skill_id: str, skill_name: str,
                  rework_count: int, session_id: str) -> None:
    """Record that output required rework after skill was active."""
    if not is_enabled():
        return
    score = max(0.0, round(1.0 - rework_count * 0.25, 4))
    conn  = _get_db()
    conn.execute(
        "INSERT INTO signals(skill_id,skill_name,session_id,kind,rework_count,score,created_at)"
        " VALUES(?,?,?,?,?,?,?)",
        (skill_id, skill_name, session_id, "rework", rework_count, score,
         datetime.now().isoformat())
    )
    conn.commit()
    conn.close()


def record_learned(skill_id: str, skill_name: str, session_id: str) -> None:
    """Record that a new skill was learned/created this session."""
    if not is_enabled():
        return
    conn = _get_db()
    conn.execute(
        "INSERT INTO signals (skill_id, skill_name, session_id, kind, score, created_at) "
        "VALUES (?,?,?,?,?,?)",
        (skill_id, skill_name, session_id, "learned", 1.0,
         datetime.now().isoformat())
    )
    conn.commit()
    conn.close()


# ---------------------------------------------------------------------------
# Analytics
# ---------------------------------------------------------------------------

def signal_stats() -> dict:
    """Overall signal stats for display."""
    if not is_enabled():
        return {"enabled": False}
    try:
        conn  = _get_db()
        total = conn.execute("SELECT COUNT(*) FROM signals").fetchone()[0]
        accepts = conn.execute(
            "SELECT COUNT(*) FROM signals WHERE kind='accept'"
        ).fetchone()[0]
        reworks = conn.execute(
            "SELECT COUNT(*) FROM signals WHERE kind='rework'"
        ).fetchone()[0]
        distinct_skills = conn.execute(
            "SELECT COUNT(DISTINCT skill_id) FROM signals"
        ).fetchone()[0]
        # Per-kind breakdown for detailed display
        kind_rows = conn.execute(
            "SELECT kind, COUNT(*) as c, AVG(score) as avg_s FROM signals GROUP BY kind"
        ).fetchall()
        conn.close()
        by_kind = {
            r[0]: {"count": r[1], "avg_score": round(r[2], 3)}
            for r in kind_rows
        }
        return {
            "enabled":         True,
            "total":           total,
            "accepts":         accepts,
            "reworks":         reworks,
            "distinct_skills": distinct_skills,
            "accept_rate":     round(accepts / total * 100, 1) if total else 0,
            "by_kind":         by_kind,
        }
    except Exception as e:
        return {"enabled": True, "error": str(e)}


def top_performing_skills(limit: int = 20) -> list[dict]:
    """
    Skills ranked by avg_score from signal data.
    Minimum 3 fires to appear (avoids noise).
    """
    if not is_enabled():
        return []
    try:
        conn = _get_db()
        rows = conn.execute(
            "SELECT skill_name, COUNT(*) as fires, "
            "AVG(score) as avg_score, "
            "SUM(CASE WHEN kind='accept' THEN 1 ELSE 0 END) as accepts, "
            "SUM(CASE WHEN kind='rework' THEN 1 ELSE 0 END) as reworks "
            "FROM signals "
            "GROUP BY skill_name "
            "HAVING fires >= 3 "
            "ORDER BY avg_score DESC, fires DESC "
            "LIMIT ?",
            (limit,)
        ).fetchall()
        conn.close()
        result = []
        for r in rows:
            d = dict(r)
            d["accept_rate"] = (
                round(d["accepts"] / d["fires"] * 100, 1)
                if d["fires"] else 0
            )
            result.append(d)
        return result
    except Exception:
        return []


# ---------------------------------------------------------------------------
# Rework detection
# ---------------------------------------------------------------------------

REWORK_WORDS = [
    "actually", "change that", "fix that", "not quite", "redo",
    "that is wrong", "no wait", "instead", "try again", "wrong",
    "that's not right", "incorrect", "nope", "revert",
]

def count_rework_signals(text: str) -> int:
    """Count rework-intent words in user text."""
    t = text.lower()
    return sum(1 for w in REWORK_WORDS if w in t)


# ---------------------------------------------------------------------------
# Future: API flush (activates when SKILLGOD_API env var is set)
# ---------------------------------------------------------------------------

def flush_signals() -> int:
    """
    Send buffered signals to Railway API.
    Noop until SKILLGOD_API env var is configured.
    Returns count of signals flushed.
    """
    api_url = os.environ.get("SKILLGOD_API", "")
    if not api_url:
        return 0  # local-only mode
    # TODO v1.1: batch POST to {api_url}/v1/signals
    return 0


# ---------------------------------------------------------------------------
# CLI
# ---------------------------------------------------------------------------

if __name__ == "__main__":
    import sys
    cmd = sys.argv[1] if len(sys.argv) > 1 else "stats"

    if cmd == "enable":
        enable()
        print("Signal tracking enabled.")
    elif cmd == "disable":
        disable()
        print("Signal tracking disabled.")
    elif cmd == "stats":
        s = signal_stats()
        if not s.get("enabled"):
            print("Signals disabled. Run: python engine/signals.py enable")
        else:
            print(f"Total signals : {s['total']}")
            print(f"Accept rate   : {s['accept_rate']}%")
            print(f"Skills tracked: {s['distinct_skills']}")
    elif cmd == "top":
        top = top_performing_skills()
        if not top:
            print("No signal data yet (min 3 fires per skill).")
        else:
            print("Top performing skills:")
            for s in top:
                print(f"  {s['skill_name']:<35} "
                      f"fires={s['fires']}  "
                      f"accept={s['accept_rate']}%  "
                      f"avg={s['avg_score']:.2f}")

File: cli/cmd/signals.go
package cmd

import (
	"fmt"
	"strings"

	"github.com/fatih/color"
	"github.com/spf13/cobra"
)

var signalsCmd = &cobra.Command{
	Use:   "signals",
	Short: "Show local signal stats and top performing skills",
	Long: `Layer 2: local signal tracking.

Signals record whether skill injections led to good outcomes (accept)
or required rework. All data stays on your machine.

  sg signals          Show stats + top performing skills
  sg signals enable   Opt in to signal tracking (default off)
  sg signals disable  Turn off signal tracking`,
	RunE: runSignals,
}

var enableSignalsCmd = &cobra.Command{
	Use:   "enable",
	Short: "Enable signal tracking",
	RunE: func(cmd *cobra.Command, args []string) error {
		sgRoot, err := findSkillGodRoot()
		if err != nil {
			return err
		}
		green := color.New(color.FgGreen).SprintFunc()
		_, err = runPython(sgRoot, "from signals import enable; enable(); print('ok')")
		if err != nil {
			return fmt.Errorf("could not enable signals: %w", err)
		}
		fmt.Printf("%s Signal tracking enabled.\n", green("[OK]"))
		fmt.Println("Signals record accept/rework outcomes per skill.")
		fmt.Println("All data stays local. Run 'sg signals' to see stats.")
		return nil
	},
}

var disableSignalsCmd = &cobra.Command{
	Use:   "disable",
	Short: "Disable signal tracking",
	RunE: func(cmd *cobra.Command, args []string) error {
		sgRoot, err := findSkillGodRoot()
		if err != nil {
			return err
		}
		yellow := color.New(color.FgYellow).SprintFunc()
		_, err = runPython(sgRoot, "from signals import disable; disable(); print('ok')")
		if err != nil {
			return fmt.Errorf("could not disable signals: %w", err)
		}
		fmt.Printf("%s Signal tracking disabled.\n", yellow("[OFF]"))
		return nil
	},
}

func init() {
	signalsCmd.AddCommand(enableSignalsCmd)
	signalsCmd.AddCommand(disableSignalsCmd)
}

func runSignals(cmd *cobra.Command, args []string) error {
	sgRoot, err := findSkillGodRoot()
	if err != nil {
		return err
	}

	bold   := color.New(color.Bold).SprintFunc()
	green  := color.New(color.FgGreen).SprintFunc()
	yellow := color.New(color.FgYellow).SprintFunc()
	red    := color.New(color.FgRed).SprintFunc()

	// Get stats
	statsCode := `
from signals import signal_stats, top_performing_skills, is_enabled
import json
print(json.dumps({
    "stats": signal_stats(),
    "top":   top_performing_skills(limit=10),
    "enabled": is_enabled(),
}))
`
	out, err := runPython(sgRoot, strings.ReplaceAll(statsCode, "\n", " "))
	if err != nil {
		return fmt.Errorf("signals engine error: %w", err)
	}

	// Simple line-based parsing (avoid JSON import in Go for now)
	out = strings.TrimSpace(out)
	if strings.Contains(out, `"enabled": false`) || strings.Contains(out, `"enabled":false`) {
		fmt.Println(bold("\nSkillGod Signals"))
		fmt.Println("────────────────────────")
		fmt.Printf("  Status: %s (opt-in required)\n", yellow("DISABLED"))
		fmt.Println()
		fmt.Printf("  Enable with: %s\n", bold("sg signals enable"))
		fmt.Println("  Signals track which skills lead to accepted vs reworked outputs.")
		fmt.Println("  All data stays local — never sent anywhere.")
		return nil
	}

	// Enabled — show stats
	fmt.Println(bold("\nSkillGod Signals"))
	fmt.Println("────────────────────────")

	// Extract key values
	getVal := func(key, s string) string {
		idx := strings.Index(s, `"`+key+`":`)
		if idx == -1 {
			return "?"
		}
		rest := s[idx+len(key)+3:]
		end  := strings.IndexAny(rest, ",}")
		if end == -1 {
			return strings.TrimSpace(rest)
		}
		return strings.TrimSpace(rest[:end])
	}

	total    := getVal("total", out)
	accepts  := getVal("accepts", out)
	reworks  := getVal("reworks", out)
	rate     := getVal("accept_rate", out)
	distinct := getVal("distinct_skills", out)

	fmt.Printf("  Total signals   : %s\n", green(total))
	fmt.Printf("  Accept rate     : %s%%\n", green(rate))
	fmt.Printf("  Accepts/Reworks : %s / %s\n", green(accepts), red(reworks))
	fmt.Printf("  Skills tracked  : %s\n", green(distinct))

	// Top skills section
	if strings.Contains(out, `"skill_name"`) {
		fmt.Println()
		fmt.Println(bold("  Top performing skills:"))
		lines := strings.Split(out, `"skill_name"`)
		for i, chunk := range lines[1:] {
			if i >= 10 {
				break
			}
			nameEnd := strings.Index(chunk, `"`)
			if nameEnd < 2 {
				continue
			}
			name := strings.Trim(chunk[:nameEnd], `: "`)
			fires := getVal("fires", chunk)
			avg   := getVal("avg_score", chunk)
			arate := getVal("accept_rate", chunk)
			fmt.Printf("    %-35s fires=%-4s accept=%s%%  avg=%s\n",
				name, fires, arate, avg)
		}
	} else {
		fmt.Println()
		fmt.Println(yellow("  No skill data yet (min 3 fires per skill to appear)."))
	}

	fmt.Println()
	fmt.Printf("  All data is local. API sync activates when %s is set.\n",
		bold("SKILLGOD_API"))
	return nil
}
